jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/insights/QueryActor.kt

```diff
@@ -0,0 +1,189 @@
+
+    private suspend fun tableLoading() = withContext(edtContext) {
+        table.setPaintBusy(true)
+        table.emptyText.text = message("loading_resource.loading")
+    }
+
+    private suspend fun tableDoneLoading() = withContext(edtContext) {
+        table.tableViewModel.fireTableDataChanged()
+        table.setPaintBusy(false)
+    }
+
+    override fun dispose() {
+        channel.close()
+    }
+    companion object {
+        private val LOG = getLogger<QueryActor<*>>()
+    }
+}
+
+class QueryResultsActor(
+    project: Project,
+    client: CloudWatchLogsClient,
+    table: TableView<MutableMap<String, String>>,
+    private val queryId: String
+) : QueryActor<MutableMap<String, String>>(project, client, table) {
+    override val emptyText = message("cloudwatch.logs.no_results_found")
+    override val tableErrorMessage = message("cloudwatch.logs.query_results_table_error")
+    override val notFoundText = message("cloudwatch.logs.no_results_found")
+
+    override suspend fun loadInitialQueryResults() {
+        var request = GetQueryResultsRequest.builder().queryId(queryId).build()
+        var response = client.getQueryResults(request)
+        while (response.results().size == 0) {
+            request = GetQueryResultsRequest.builder().queryId(queryId).build()
+            response = client.getQueryResults(request)
+        }
+        val listOfResults = mutableListOf<MutableMap<String, String>>()
+        val queryResults = response.results().filterNotNull()
+        for (result in queryResults) {
+            val fieldToValueMap = mutableMapOf<String, String>()
+            for (field in result) {
+                fieldToValueMap[field.field().toString()] = field.value().toString()
+                if (field.field() == "@ptr") {
+                    queryResultsIdentifierList.add(field.value())
+                }
+            }
+            listOfResults.add(fieldToValueMap)
+        }
+        moreResultsAvailable = response.statusAsString() != "Complete"
+        loadAndPopulateResultsTable { listOfResults }
+    }
+
+    override suspend fun loadNext(): List<MutableMap<String, String>> {
+        val request = GetQueryResultsRequest.builder().queryId(queryId).build()
+        val response = client.getQueryResults(request)
+        moreResultsAvailable = response.statusAsString() != "Complete"
+        return checkIfNewResult(response.results().filterNotNull())
+    }
+
+    fun checkIfNewResult(queryResultList: List<MutableList<ResultField>>): List<MutableMap<String, String>> {
+        val listOfResults = mutableListOf<MutableMap<String, String>>()
+        for (result in queryResultList) {
+            var resultIsUnique = false
+            var fieldToValueMap = mutableMapOf<String, String>()
+            for (field in result) {
+                fieldToValueMap[field.field().toString()] = field.value().toString()
+                if (field.field() == "@ptr" && field.value() !in queryResultsIdentifierList) {
```

Comment:
the const `"@ptr"` is suspicious, can you add a comment why it's here?

Reply:
ptr@ is a unique identifier for each resultant log event used here to verify if this result has already been included in the table